From 116cf14899e2fa905f47a99abfa6c0ef2783103d Mon Sep 17 00:00:00 2001
From: Colin Ian King <colin.king@canonical.com>
Date: Mon, 10 Mar 2014 17:11:45 +0000
Subject: [PATCH 28/28] get_preference_cstr: return C string allocated on heap

The original code returned a C string that was effectively
destroyed when exiting the scope.  Best to dup this C string
so it is on the heap and use and free it by the caller.

Signed-off-by: Colin Ian King <colin.king@canonical.com>
---
 src/main.cpp           | 5 +++++
 src/thd_preference.cpp | 2 +-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/main.cpp b/src/main.cpp
index 5e1b369..0eda788 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -140,7 +140,12 @@ gboolean thd_dbus_interface_get_current_preference(PrefObject *obj,
 
 	cthd_preference thd_pref;
 	value_out = (gchar*) thd_pref.get_preference_cstr();
+	if (!value_out) {
+		g_free(pref_str);
+		return FALSE;
+	}
 	strncpy(pref_str, value_out, MAX_DBUS_REPLY_STR_LEN);
+	free(value_out);
 	thd_log_debug("thd_dbus_interface_get_current_preference out :%s\n",
 			pref_str);
 	*pref_out = pref_str;
diff --git a/src/thd_preference.cpp b/src/thd_preference.cpp
index 561a71a..cfcc5e8 100644
--- a/src/thd_preference.cpp
+++ b/src/thd_preference.cpp
@@ -78,7 +78,7 @@ std::string cthd_preference::get_preference_str() {
 }
 
 const char *cthd_preference::get_preference_cstr() {
-	return int_pref_to_string(preference).c_str();
+	return strdup(int_pref_to_string(preference).c_str());
 }
 
 int cthd_preference::get_preference() {
-- 
1.9.0

